chronos/asyncproc.nim

```diff
@@ -0,0 +1,1323 @@
+        flags = baseFlags + {StreamHolderFlag.Stream,
+                             StreamHolderFlag.Transport}
+      ok(AsyncStreamHolder(kind: StreamKind.Reader, reader: reader,
+                           flags: flags))
+    of StreamKind.Writer:
+      let
+        transp = ? fromPipe2(handle.handle)
+        writer = newAsyncStreamWriter(transp)
+        flags = baseFlags + {StreamHolderFlag.Stream,
+                             StreamHolderFlag.Transport}
+      ok(AsyncStreamHolder(kind: StreamKind.Writer, writer: writer,
+                           flags: flags))
+    of StreamKind.None:
+      ok(AsyncStreamHolder(kind: StreamKind.None))
+  of ProcessStreamHandleKind.Transport:
+    case kind
+    of StreamKind.Reader:
+      let
+        reader = newAsyncStreamReader(handle.transp)
+        flags = baseFlags + {StreamHolderFlag.Stream}
+      ok(AsyncStreamHolder(kind: StreamKind.Reader, reader: reader,
+                           flags: flags))
+    of StreamKind.Writer:
+      let
+        writer = newAsyncStreamWriter(handle.transp)
+        flags = baseFlags + {StreamHolderFlag.Stream}
+      ok(AsyncStreamHolder(kind: StreamKind.Writer, writer: writer,
+                           flags: flags))
+    of StreamKind.None:
+      ok(AsyncStreamHolder(kind: StreamKind.None))
+  of ProcessStreamHandleKind.StreamReader:
+    ok(AsyncStreamHolder(kind: StreamKind.Reader, reader: handle.reader,
+                         flags: baseFlags))
+  of ProcessStreamHandleKind.StreamWriter:
+    ok(AsyncStreamHolder(kind: StreamKind.Writer, writer: handle.writer,
+                         flags: baseFlags))
+  of ProcessStreamHandleKind.None, ProcessStreamHandleKind.Auto:
+    ok(AsyncStreamHolder(kind: StreamKind.None))
+
+proc init*(t: typedesc[ProcessStreamHandle]): ProcessStreamHandle =
+  ProcessStreamHandle(kind: ProcessStreamHandleKind.None)
+
+proc init*(t: typedesc[ProcessStreamHandle],
+           handle: AsyncFD): ProcessStreamHandle =
+  ProcessStreamHandle(kind: ProcessStreamHandleKind.ProcHandle, handle: handle)
+
+proc init*(t: typedesc[ProcessStreamHandle],
+           transp: StreamTransport): ProcessStreamHandle =
+  doAssert(transp.kind == TransportKind.Pipe,
+           "Only pipe transports can be used as process streams")
+  ProcessStreamHandle(kind: ProcessStreamHandleKind.ProcHandle, transp: transp)
+
+proc init*(t: typedesc[ProcessStreamHandle],
+           reader: AsyncStreamReader): ProcessStreamHandle =
+  ProcessStreamHandle(kind: ProcessStreamHandleKind.StreamReader,
+                      reader: reader)
+
+proc init*(t: typedesc[ProcessStreamHandle],
+           writer: AsyncStreamWriter): ProcessStreamHandle =
+  ProcessStreamHandle(kind: ProcessStreamHandleKind.StreamWriter,
+                      writer: writer)
+
+proc isEmpty*(handle: ProcessStreamHandle): bool =
+  handle.kind == ProcessStreamHandleKind.None
+
+proc suspend*(p: AsyncProcessRef): AsyncProcessResult[void] {.gcsafe.}
```

Comment:
procs with forward declarations will be assumed to `raises: [Exception]`, so consider adding explicit `raises:` annotation here and elsewhere.

Reply:
Fixed.